### ISSUE TITLE
Incorrect rowVars()/colVars() results when 'center' is specified

### ISSUE THREAD

**Issue by hpages** [HenrikBengtsson/matrixStats#183]
Hi Henrik,

This doesn't look correct:
```
library(matrixStats)
m <- matrix(1:24, ncol=6)
matrixStats::colVars(m, cols=1:2, center=6)
# [1] -38  NA
```
The correct answer being probably:
```
matrixStats::colVars(m[ , 1:2], center=6)
# [1] -38  10
```

Also this doesn't look correct either:
```
a <- array(1:24, 4:2)
matrixStats::colVars(a, center=6, dim.=dim(m))
#          [,1]     [,2]
# [1,] -38.0000 234.0000
# [2,]  10.0000 410.0000
# [3,] 100.6667 628.6667

matrixStats::rowVars(a, center=6, dim.=dim(m))
# [1] 197.5 232.0 269.5 310.0
```
The correct answers being probably:
```
matrixStats::colVars(m, center=6)
# [1] -38.0000  10.0000 100.6667 234.0000 410.0000 628.6667

matrixStats::rowVars(m, center=6)
# [1] 158.0 185.6 215.6 248.0
```

Finally the two calls below should probably return an error:
```
matrixStats::colVars(a, center=6)
#          [,1]     [,2]
# [1,] -38.0000 234.0000
# [2,]  10.0000 410.0000
# [3,] 100.6667 628.6667

matrixStats::rowVars(a, center=6)
# [1] 197.5 232.0 269.5 310.0
```

**sessionInfo():**
```
R version 4.0.2 Patched (2020-06-24 r78746)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.7 LTS

Matrix products: default
BLAS:   /home/hpages/R/R-4.0.r78746/lib/libRblas.so
LAPACK: /home/hpages/R/R-4.0.r78746/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] matrixStats_0.57.0

loaded via a namespace (and not attached):
[1] compiler_4.0.2 tools_4.0.2   
```

**Comment by karoliskoncevicius** [comment]
This:

    matrixStats::colVars(m[ , 1:2], center=6)
    # [1] -38  10

Cannot be correct either as variance cannot be negative.

**Comment by karoliskoncevicius** [comment]
This affects some of my stuff too, so I took a quick look at the code. Seems like the function expects `center` to be one value per column, not just a single number: [rowVars line 115](https://github.com/HenrikBengtsson/matrixStats/blob/develop/R/rowVars.R#L115). As you can see it subsets the `center` by `cols`, so ends up calling `6[1:2]`, which explains the NA in the second position of your example.

The negative variance is still worrisome, unless I misunderstood what `center` is.

**Comment by HenrikBengtsson** [comment]
Thanks for reporting. Turns out the problem is lack of documentation and validation.  I've added the following assertion:

```r
library(matrixStats)
m <- matrix(1:24, ncol=6)
matrixStats::colVars(m, cols=1:2, center=6)
## Error in matrixStats::colVars(m, cols = 1:2, center = 6) : 
##   Argument 'center' should be of the same length as number of rows of 'x': 1 != 6
```

Also for

```r
library(matrixStats)
m <- matrix(1:24, ncol=6)
matrixStats::colVars(m, center=6)
## Error in matrixStats::colVars(m, center = 6) : 
##   Argument 'center' should be of the same length as number of rows of 'x': 1 != 6
```

The rationale is that the default corresponds to:

```r
library(matrixStats)
m <- matrix(1:24, ncol=6)
center <- colMeans(m)
matrixStats::colVars(m, center=center)
## [1] 1.666667 1.666667 1.666667 1.666667 1.666667 1.666667
apply(m - 6, MARGIN = 2, FUN = var)
[1] 1.666667 1.666667 1.666667 1.666667 1.666667 1.666667
```

and

```r
library(matrixStats)
m <- matrix(1:24, ncol=6)
center <- colMeans(m)
matrixStats::colVars(m, cols=1:2, center=center)
## [1] 1.666667 1.666667
## apply(m[,1:2] - 6, MARGIN = 2, FUN = var)
[1] 1.666667 1.666667
```


# Actions

I'll probably have to check and add this asserting to other functions that take an argument `center':

* [x] `colVars()` and `rowVars()`
* [x] `colSds()` and `rowSds()`
* [ ] `colMads()` and `rowMads()`

**Comment by HenrikBengtsson** [comment]
The above comment is regarding the wrong length of `center`.  The negative estimate, e.g.

```r
> m <- matrix(1:24, ncol=6)
> matrixStats::colVars(m, center=rep(6, times=ncol(m)))
[1] -38.0000  10.0000 100.6667 234.0000 410.0000 628.6667
```

is because the `center`-adjusted variance is calculated as:

Eqn (1):
```r
    n <- nrow(x)
    x <- x * x
    x <- colMeans(x, na.rm = na.rm)
    x <- (x - center^2)
    x * (n/(n - 1))
```
which gives:
```r
[1] -38.0000  10.0000 100.6667 234.0000 410.0000 628.6667
```
This calculation is based on the assumption that **`center` is a proper estimate of the column means**.  

To support any type of `center` values, that is, `center` specifies how much the matrix values should be **translated**, then we'd have to calculate it as:

Eqn (2):
```r
    n <- nrow(x)
    x <- x - center
    x <- x * x
    x <- colMeans(x, na.rm = na.rm)
    x * (n/(n - 1))
```
to get:
```r
[1]  18.00000   2.00000  28.66667  98.00000 210.00000 364.66667
```

I don't recall the reason for Eqn (1) vs Eqn (2), but I suspect that it's because Eqn (1) was faster or more memory efficient.  It it also not unlikely Eqn (1) was the only implementation before a native-code implementation was provided and `center` was exposed to the user as an argument.  At that time, Eqn (1) was guaranteed to be correct.

Looking at `colMads()`, there we use the `x <- x - center` strategy, which obviously is because that's the only alternative.

# Action

* [ ] Update help that `center` must be "proper" estimates of the column/row means in order to get "proper" variance estimates of matrix `x` (sic!).  This is true regardless of Eqn (1) or Eqn (2) being used.  Eqn (2) is only correct if we're interested in the variance of translated `x - center` - a use that should be discouraged
* [ ] Profile Eqn (1) vs Eqn (2) for speed and memory
* [ ] If no obvious different, switch to Eqn (2) immediately to protect again `center` estimate mistakes (some which might be out there already)


Comments?

**Comment by karoliskoncevicius** [comment]
I think using center solely for providing correct means, to speed up computing, is good. That avoids the negative variance scenarios, expect when caused by user error.

But in that case maybe it would be better to start or end the argument name with a dot? i.e. `center.`, just like `dim.` ? If I understood it correctly that the dot indicates a special optional argument. It probably wouldn't even break backwards compatibility as function arguments are auto-completed.

**Comment by HenrikBengtsson** [comment]
> But in that case maybe it would be better to start or end the argument name with a dot? i.e. center., just like dim. ? If I understood it correctly that the dot indicates a special optional argument. ...

It's actually because we can't use `dim = dim(x)`, cf. https://github.com/HenrikBengtsson/matrixStats/issues/163.  But this should be explained in the help:

* [ ] Update help to explain why the argument is named `dim.` and not just `dim`.

Having said this, it might not be a bad idea to make these types of low-level arguments less pronounced so that they are not used incorrectly by mistake.

**Comment by HenrikBengtsson** [comment]
@hpages, https://github.com/Bioconductor/MatrixGenerics/blob/master/tests/testthat/test-api_compatibility.R needs to be updated.   I've updated develop version of matrixStats to give an error if `length(center)` does not match the number of rows/columns in the matrix.  I've also allowed for `length(center) == 1L` in the case where `rows` / `cols` is _not_ specific - support for scalar `center` is deprecated but still works because it does and has always given a correct value.

**Comment by HenrikBengtsson** [comment]
The problem regarding `matrixStats::rowVars(a, ..., dim.=dim(m))` has been moved to https://github.com/HenrikBengtsson/matrixStats/issues/185

**Comment by const-ae** [comment]
Is there a reason that now `colWeightedMad` throws an [error](https://github.com/HenrikBengtsson/matrixStats/blob/develop/R/weightedMad.R#L160) `if (length(center) > 1L)`, whereas `colMads` throws an error if  `length(center) != ncol(x)`?

**Comment by HenrikBengtsson** [comment]
> Is there a reason that now colWeightedMad throws an error if (length(center) > 1L), whereas colMads throws an error if length(center) != ncol(x)?

No good reason.  I suspect the reason is that `center` was left as an argument in a very conservative manner.  It could also be a conservative cut'n'paste from `weightedMad()`, which existed way before the matrix implementations.  Similarly to name attributes, argument `center` is definitely up there among the things that never good enough love/deep focus.

# Action

* [ ] Add support for `length(center) == ncol(x)` to `colWeightedMads()` and `rowWeightedMads()` with deprecated backward-compatibility support for length(center) == 1L`.   Comment: Scalar `center` doesn't really make sense.  It would make sense if we know that the mean parameters are all identical (== `center`).  However, using _known_ parameters requires another degree-of-freedom correction, which we don't support.  In other words, `center` should only be _estimates_ and with estimates it is extremely^6 unlikely they are all identical.  Thus, a scalar `center` to be used for all columns/rows, does not make much sense.

**Comment by jan-glx** [comment]
I believe this has been raised before in #160 and with a proposed fix in #161.

**Comment by HenrikBengtsson** [comment]
The remaining tasks here were:

* [ ] Profile Eqn (1) vs Eqn (2) for speed and memory
* [ ] If no obvious difference, switch to Eqn (2) immediately to protect again center estimate mistakes (some which might be out there already)

However, these are now moot, since correctness takes precedence over speed.  If performance is worse, it should be address in native code anyway.

The remaining parts of this issue is now tracked in Issue #160.  Closing this one.
